src/reseeding.rs

```diff
@@ -39,73 +40,127 @@ use {RngCore, SeedableRng, Error, ErrorKind};
 ///   strong and with no known attack, does not come with any proof of security
 ///   and does not meet the current standards for a cryptographically secure
-///   PRNG. By reseeding it frequently (every 32 MiB) it seems safe to assume
+///   PRNG. By reseeding it frequently (every 32 kiB) it seems safe to assume
 ///   there is no attack that can operate on the tiny window between reseeds.
 ///
 /// # Error handling
 ///
 /// If reseeding fails, `try_fill_bytes` is the only `Rng` method to report it.
 /// For all other `Rng` methods, `ReseedingRng` will not panic but try to
-/// handle the error intelligently; if handling the source error fails these
+/// handle the error intelligently through some combination of retrying and
+/// delaying reseeding until later. If handling the source error fails these
 /// methods will continue generating data from the wrapped PRNG without
 /// reseeding.
-///
-/// It is usually best to use the infallible methods `next_u32`, `next_u64` and
-/// `fill_bytes` because they can make use of this error handling strategy.
-/// Use `try_fill_bytes` and possibly `try_reseed` if you want to handle
-/// reseeding errors explicitly.
 #[derive(Debug)]
-pub struct ReseedingRng<R, Rsdr> {
-    rng: R,
-    reseeder: Rsdr,
-    threshold: i64,
-    bytes_until_reseed: i64,
-}
+pub struct ReseedingRng<R, Rsdr>(BlockRng<Reseeder<R, Rsdr>>)
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore;
 
-impl<R: RngCore + SeedableRng, Rsdr: RngCore> ReseedingRng<R, Rsdr> {
+impl<R, Rsdr> ReseedingRng<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
     /// Create a new `ReseedingRng` with the given parameters.
     ///
     /// # Arguments
     ///
     /// * `rng`: the random number generator to use.
     /// * `threshold`: the number of generated bytes after which to reseed the RNG.
     /// * `reseeder`: the RNG to use for reseeding.
-    pub fn new(rng: R, threshold: u64, reseeder: Rsdr) -> ReseedingRng<R,Rsdr> {
+    pub fn new(rng: R, threshold: u64, reseeder: Rsdr)
+        -> ReseedingRng<R, Rsdr>
+    {
         assert!(threshold <= ::core::i64::MAX as u64);
-        ReseedingRng {
-            rng: rng,
-            reseeder: reseeder,
-            threshold: threshold as i64,
-            bytes_until_reseed: threshold as i64,
+        let results_empty = R::Results::default();
+        ReseedingRng(
+            BlockRng {
+                core: Reseeder {
+                    core: rng,
+                    reseeder: reseeder,
+                    threshold: threshold as i64,
+                    bytes_until_reseed: threshold as i64,
+                },
+                index: results_empty.as_ref().len(), // generate on first use
+                results: results_empty,
+            }
+        )
+    }
+
+    /// Reseed the internal PRNG.
+    pub fn reseed(&mut self) -> Result<(), Error> {
+        self.0.core.reseed()
```

Comment:
This doesn't adjust `bytes_until_reseed`; it should probably call `auto_reseed` instead...

Reply:
This is done on purpose, see the previous comment

Reply:
But if reseeding *is* successful, then it makes sense to reset `bytes_until_reseed` IMO.